backend/canisters/local_user_index/impl/src/model/referral_codes.rs

```diff
@@ -43,51 +45,75 @@ pub struct ReferralTypeMetrics {
     pub total: usize,
 }
 
+pub enum ReferralCodeError {
+    NotFound,
+    AlreadyClaimed,
+    Expired,
+}
+
 impl ReferralCodes {
-    pub fn add(&mut self, referral_type: ReferralType, code: String, now: TimestampMillis) -> bool {
+    pub fn set_miami_codes_expired(&mut self, now: TimestampMillis) {
+        for code in self.codes.values_mut() {
+            if matches!(code.referral_type, ReferralType::BtcMiami) && code.expiry.is_none() {
+                code.expiry = Some(now);
+            }
+        }
+    }
+
+    pub fn add(
+        &mut self,
+        referral_type: ReferralType,
+        code: String,
+        expiry: Option<TimestampMillis>,
+        now: TimestampMillis,
+    ) -> bool {
         match self.codes.entry(code) {
             Entry::Occupied(_) => false,
             Entry::Vacant(e) => {
                 e.insert(ReferralCodeDetails {
                     referral_type,
                     created: now,
                     claimed: None,
+                    expiry,
                 });
                 true
             }
         }
     }
 
-    pub fn claim(&mut self, code: String, user_id: UserId, now: TimestampMillis) -> bool {
-        match self.codes.entry(code) {
-            Entry::Occupied(mut e) => {
-                let details = e.get_mut();
-                match details.claimed {
-                    Some(_) => false,
-                    None => {
-                        details.claimed = Some(ReferralCodeClaim { when: now, who: user_id });
-                        true
-                    }
-                }
-            }
-            Entry::Vacant(_) => false,
-        }
-    }
+    // pub fn claim(&mut self, code: String, user_id: UserId, now: TimestampMillis) -> bool {
```

Comment:
Why is this commented out?

Reply:
I'll just put an `allow(dead_code)` over it.

Reply:
Actually I've deleted it instead. We can recover it if and when it is needed.